realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -15,6 +15,84 @@ using namespace ddynamic_reconfigure;
+
+        for (int y = 0; y < height; ++y)
+        {
+            const uint8_t *p_src = src + (y * width);
+            int *p_sum = sum_x + (y * width);
+
+            p_sum[0] = r * p_src[0];
+            for (int xx = 0; xx <= r; ++xx) p_sum[0] += int(p_src[xx]);
+
+            for (int x = 1;         x < r + 1;     ++x) p_sum[x] = p_sum[x - 1] + int(p_src[x + r]) - int(p_src[0]);
+            for (int x = r + 1;     x < width - r; ++x) p_sum[x] = p_sum[x - 1] + int(p_src[x + r]) - int(p_src[x - r - 1]);
+            for (int x = width - r; x < width;     ++x) p_sum[x] = p_sum[x - 1] + int(p_src[width - 1]) - int(p_src[x - r - 1]);
+        }
+        { // y = 0
+            for (int x = 0; x < width; ++x) {
+                int *p_src = sum_x + x;
+                int *p_sum = sum_xy + x;
+
+                p_sum[0] = r * p_src[0];
+                for (int yy = 0; yy <= r; ++yy) p_sum[0] += int(p_src[yy * width]);
+            }
+        }
+        for (int y = 1; y < r + 1; ++y)
+        {
+            for (int x = 0; x < width; ++x) {
+                int *p_src = sum_x + x;
+                int *p_sum = sum_xy + x;
+                p_sum[y * width] = p_sum[(y - 1) * width] + int(p_src[(y + r) * width]) - int(p_src[0 * width]);
+            }
+        }
+        for (int y = r + 1; y < height - r; ++y)
+        {
+            for (int x = 0; x < width; ++x) {
+                int *p_src = sum_x + x;
+                int *p_sum = sum_xy + x;
+                p_sum[y * width] = p_sum[(y - 1) * width] + int(p_src[(y + r) * width]) - int(p_src[(y - r - 1) * width]);
+            }
+        }
+        for (int y = height - r; y < height; ++y)
+        {
+            for (int x = 0; x < width; ++x) {
+                int *p_src = sum_x + x;
+                int *p_sum = sum_xy + x;
+                p_sum[y * width] = p_sum[(y - 1) * width] + int(p_src[(height - 1) * width]) - int(p_src[(y - r - 1) * width]);
+            }
+        }
+    }
+
+    void mark_bright_regions(const uint8_t* guide, uint8_t* bright, int width, int height, int r_ero, int r_dil, int thresh)
+    {
+        std::unique_ptr<int []> buf(new int[width * height]);
+        std::unique_ptr<int []> sum(new int[width * height]);
+        int* p_buf = buf.get();
+        int* p_sum = sum.get();
+
+        // set 1 if brighter than 'thresh'
+        for(int i = 0; i < width * height; ++i) bright[i] = (guide[i] > thresh ? 1 : 0);
+
+        // if all the pixels around are brighter than 'thresh'
+        box_filter(bright, p_sum, p_buf, width, height, r_ero); // erosion
+        int max_sum = (2 * r_ero + 1) * (2 * r_ero + 1);
+        for (int i = 0; i < width * height; ++i) bright[i] = (p_sum[i] == max_sum ? 1 : 0);
+
+        // if any of the pixels around are bright
+        box_filter(bright, p_sum, p_buf, width, height, r_dil); // dilation
+        for (int i = 0; i < width * height; ++i) bright[i] = (p_sum[i] > 0 ? 1 : 0);
```

Comment:
- I will modify the comments.
  Erosion is a technical term for computing the minimum value over a certain range, which is (2)+(3).
  So, I will change the comments to make the steps (1) - (5) to be clear

- Your algorithm seems more robust. However, single dark pixel such as an dead pixel will not be an issue because r_dil is bigger than r_ero.
  Maybe your idea is better because it is simpler? 

- Searching for (2r+1) x (2r+1) for each pixel takes O(r^2) time, whereas the box filter is near O(1) operation. Because 'r' for dilation might be very big, it is safe to use O(1) algorithm.